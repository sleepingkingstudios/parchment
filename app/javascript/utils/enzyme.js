import { valueOrDefault } from './object';
import { safeCapitalize } from './string';

export const extractInjectedComponent = (component) => {
  const injected = component.shallow().find('ComponentWithInjectedProps');

  return injected.exists() ? injected : component;
};

export const toHaveFormControl = expect => (received, expectedControl, withProps) => {
  const formControl = received.find(expectedControl);
  const expectedProps = valueOrDefault(withProps, {});

  expect(formControl).toExist();
  expect(formControl).toHaveProp(expectedProps);

  return { pass: true, message: () => '' };
};

export const toHaveFormGroup = expect => (received, expectedGroup, fn) => {
  const formGroup = received.find(expectedGroup);

  expect(formGroup).toExist();

  fn(extractInjectedComponent(formGroup).shallow());

  return { pass: true, message: () => '' };
};

export const toHaveFormInput = expect => (received, props) => {
  const {
    form,
    inputType,
    prop,
  } = props;

  expect(received).toHaveProp({ prop });

  const formInput = received.find(`${safeCapitalize(prop)}Input`);

  expect(formInput).toExist();
  expect(formInput).toHaveProp({ form });

  const shallowInput = formInput.shallow();

  expect(shallowInput).toExist();
  expect(shallowInput).toHaveDisplayName(valueOrDefault(inputType, 'FormInput'));

  return { pass: true, message: () => '' };
};
