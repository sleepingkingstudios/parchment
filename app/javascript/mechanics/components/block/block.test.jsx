import React from 'react';
import { shallow } from 'enzyme';

import MechanicBlock from './block';

describe('<MechanicBlock />', () => {
  const mechanic = {
    id: '00000000-0000-0000-0000-000000000000',
    type: '',
    name: 'Self-Destruct',
    shortDescription: 'Creates a big boom',
    description: 'Destroys all nearby foes at the cost of your life.',
  };
  const defaultProps = { data: mechanic };

  it('should wrap the contents in a <div> element', () => {
    const rendered = shallow(<MechanicBlock {...defaultProps} />);

    expect(rendered).toHaveDisplayName('div');
    expect(rendered).toHaveClassName('mechanic-block');
  });

  it('should render the mechanic name', () => {
    const rendered = shallow(<MechanicBlock {...defaultProps} />);
    const nameElement = rendered.find('.mechanic-block-name');

    expect(nameElement).toHaveDisplayName('p');
    expect(nameElement).toIncludeText(mechanic.name);
  });

  it('should render the mechanic short description', () => {
    const rendered = shallow(<MechanicBlock {...defaultProps} />);
    const shortDescriptionElement = rendered.find('.mechanic-block-short-description');

    expect(shortDescriptionElement).toHaveDisplayName('p');
    expect(shortDescriptionElement).toHaveText(`Short Description: ${mechanic.shortDescription}`);
  });

  it('should render the mechanic description', () => {
    const rendered = shallow(<MechanicBlock {...defaultProps} />);
    const descriptionElement = rendered.find('PlainText');

    expect(descriptionElement).toExist();
    expect(descriptionElement).toHaveClassName('mechanic-block-description');
    expect(descriptionElement).toHaveProp('text', mechanic.description);
  });

  describe('when the mechanic has a type', () => {
    const action = Object.assign(
      {},
      mechanic,
      { type: 'Mechanics::Action' },
    );

    it('should render the mechanic name', () => {
      const rendered = shallow(<MechanicBlock {...defaultProps} data={action} />);
      const nameElement = rendered.find('.mechanic-block-name');

      expect(nameElement).toHaveDisplayName('p');
      expect(nameElement).toIncludeText(mechanic.name);
    });

    it('should render the mechanic type', () => {
      const rendered = shallow(<MechanicBlock {...defaultProps} data={action} />);
      const nameElement = rendered.find('.mechanic-block-name');
      const typeElement = nameElement.find('MechanicBlockType').shallow();

      expect(typeElement).toHaveText('(Action)');
    });
  });
});
