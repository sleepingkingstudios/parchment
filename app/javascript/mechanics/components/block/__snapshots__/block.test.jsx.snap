// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MechanicBlock /> should match the snapshot 1`] = `
<div
  className="mechanic-block"
>
  <p
    className="mechanic-block-name"
  >
    Self-Destruct
     
    <MechanicBlockType
      description="Destroys all nearby foes at the cost of your life."
      id="00000000-0000-0000-0000-000000000000"
      name="Self-Destruct"
      shortDescription="Creates a big boom"
      slug="self-destruct"
      type=""
    />
  </p>
  <RichText
    className="mechanic-block-description"
    liquid={true}
    markdown={true}
    text="Destroys all nearby foes at the cost of your life."
  />
</div>
`;

exports[`<MechanicBlock /> with showAdditionalDetails: true should match the snapshot 1`] = `
<div
  className="mechanic-block"
>
  <p
    className="mechanic-block-name"
  >
    Self-Destruct
     
    <MechanicBlockType
      description="Destroys all nearby foes at the cost of your life."
      id="00000000-0000-0000-0000-000000000000"
      name="Self-Destruct"
      notes="It's a real blast"
      shortDescription="Creates a big boom"
      slug="self-destruct"
      type=""
    />
  </p>
  <RichText
    className="mechanic-block-description"
    liquid={true}
    markdown={true}
    text="Destroys all nearby foes at the cost of your life."
  />
  <hr />
  <div
    className="mechanic-block-additional-details"
  >
    <p
      className="mechanic-block-slug"
    >
      <em>
        Slug:
      </em>
       
      self-destruct
    </p>
    <p
      className="mechanic-block-short-description"
    >
      <em>
        Short Description:
      </em>
       
      Creates a big boom
    </p>
    <p>
      <em>
        Notes
      </em>
    </p>
    <PlainText
      className="mechanic-block-notes"
      text="It's a real blast"
    />
  </div>
</div>
`;
